cpp/src/arrow/testing/util.h

```diff
@@ -122,4 +123,66 @@ class BatchIterator : public RecordBatchReader {
   size_t position_;
 };
 
+struct MakeRepeatedArrayImpl {
+  template <typename T>
+  MakeRepeatedArrayImpl(std::shared_ptr<DataType> type, const T& value,
+                        int64_t repetitions)
+      : type_(type), typeid_(Typeid<T>()), value_(&value), repetitions_(repetitions) {}
+
+  template <typename T>
+  enable_if_number<T, Status> Visit(const T& t) {
+    typename TypeTraits<T>::BuilderType builder;
+    ARROW_ASSIGN_OR_RAISE(auto value, GetValue<typename T::c_type>());
+    RETURN_NOT_OK(builder.Resize(repetitions_));
+    for (auto i = 0; i < repetitions_; ++i) {
+      builder.UnsafeAppend(value);
+    }
+    return builder.Finish(&out_);
+  }
+
+  template <typename T>
+  enable_if_binary_like<T, Status> Visit(const T& t) {
+    typename TypeTraits<T>::BuilderType builder(type_, default_memory_pool());
+    ARROW_ASSIGN_OR_RAISE(auto value, GetValue<util::string_view>());
+    RETURN_NOT_OK(builder.Resize(repetitions_));
+    for (auto i = 0; i < repetitions_; ++i) {
+      builder.UnsafeAppend(value);
+    }
+    return builder.Finish(&out_);
+  }
+
+  Status Visit(const DataType& t) {
+    return Status::Invalid("creating arrays of repeated ", t);
+  }
+
+  template <typename T>
+  static const bool* Typeid() {
+    static const bool _ = true;
+    return &_;
+  }
+
+  template <typename T>
+  Result<T> GetValue() {
+    if (typeid_ != Typeid<T>()) {
+      return Status::TypeError("cannot create an array of type ", *type_,
+                               " from given value");
+    }
+    return *static_cast<const T*>(value_);
+  }
+
+  std::shared_ptr<DataType> type_;
+  const bool* typeid_;
+  const void* value_;
+  int64_t repetitions_;
+  std::shared_ptr<Array> out_;
+};
+
+template <typename T>
+Result<std::shared_ptr<Array>> MakeRepeatedArray(const std::shared_ptr<DataType>& type,
+                                                 int64_t repetitions, const T& t) {
+  MakeRepeatedArrayImpl impl(type, t, repetitions);
+  RETURN_NOT_OK(VisitTypeInline(*type, &impl));
+  return impl.out_;
+}
```

Comment:
This seems rather specialized. Might it be better to have a helper function like

`AppendRepeats(const T& val, int64_t num_repeats, ArrayBuilder* builder)`

?

Reply:
I could add that as a backend; it does seem reusable, but I think replacing MakeRepeatedArray with this would add some unnecessary boilerplate to our tests

Reply:
I'd say we hide this function, I created a "better" version in my next patch, see https://github.com/fsaintjacques/arrow/compare/ARROW-6161-parquet-scan...fsaintjacques:ARROW-6238-simple-datasource-datafragment#diff-cd3964185575ae51994d69c6b894dbf3R348-R410

It's a mixin that yields Array/RecordBatch/RecordBatchReader of a giving schema when you don't really care about the content.

Reply:
For now, I've replaced this with a more generic replacement. We can decide what to keep in the next PR